### PR TITLE
[jvm-packages] Move cache files to TempDirectory and delete this directory after XGBoost job finishes

### PR DESCRIPTION
**Pull request by Yunni**
As these are cache files, we can apply these changes to avoid naming conflicts and leftovers after Spark jobs finish.

### CODE REVIEW

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 1]
```diff
@@ -120,11 +121,9 @@ object XGBoost extends Serializable {
       }
       val taskId = TaskContext.getPartitionId().toString
       val cacheDirName = if (useExternalMemory) {
-        val dir = new File(s"${TaskContext.get().stageId()}-cache-$taskId")
-        if (!(dir.exists() || dir.mkdirs())) {
-          throw new XGBoostError(s"failed to create cache directory: $dir")
-        }
-        Some(dir.toString)
+        val dir = Files.createTempDirectory(s"${TaskContext.get().stageId()}-cache-$taskId")
+        new File(dir.toUri).deleteOnExit()
```
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Why not delete it manually after the task is done (i.e. in `Watches`)?

Comment (Yunni) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Make senses. Done.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 2]
```diff
@@ -480,11 +479,7 @@ private class Watches private(
   def delete(): Unit = {
     toMap.values.foreach(_.delete())
     cacheDirName.foreach { name =>
-      for (cacheFile <- new File(name).listFiles()) {
-        if (!cacheFile.delete()) {
-          throw new IllegalStateException(s"failed to delete $cacheFile")
-        }
-      }
+      FileUtils.deleteDirectory(new File(name))
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: what's the difference with the previous approach then?

Comment (Yunni) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: The previous approach leaves an empty directory.

Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Do we already depend on commons-io?

Comment (Yunni) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Yes, both Spark and Flink depends on `commons-io`.

**Comment by superbobry** [comment]
@Yunni can you update the PR title/description to reflect the current implementation?

**Comment by Yunni** [comment]
Looks like there is issue with travis. Re-trigger the build.

**Comment by superbobry** [comment]
The diff looks strange, why the C++ changes?

**Comment by Yunni** [comment]
Ops, Looks like I messed up the commits

**Comment by CodingCat** [comment]
what happened on this?

**Comment by Yunni** [comment]
Nothing big. I did not merge successfully so some other commits on master branch was shown in this PR.

Travis image seems to be very slow.

**Comment by CodingCat** [comment]
all tests with mac os x down

**Comment by Yunni** [comment]
@CodingCat They posted an outage these two days: https://www.traviscistatus.com/incidents/mc9x2wmpnvhg

**Comment by CodingCat** [comment]
ok, then let's wait for their fix

**Comment by obones** [comment]
Has this been tested under Windows x64?
I'm getting crashes that I suspect are related, as mentioned here: https://github.com/dmlc/xgboost/issues/3038#issuecomment-359403745

**Comment by superbobry** [comment]
It is tested on Windows x64: https://ci.appveyor.com/project/tqchen/xgboost/build/1.0.1553/job/o7ajxln6ybvwa8dg

**Comment by obones** [comment]
Thanks, I was sure there were unit tests running, and so you do not get crashes.
But are the files created at the proper location? I'm asking because this is not what I'm seeing in the mentioned issue and I can't see a reason for how it behaves

**Pull request merged**